io_context.go

```diff
@@ -25,15 +30,36 @@ func (ic *IOContext) Closep() error {
 	return newError(C.avio_closep(&ic.c))
 }
 
+func (ic *IOContext) Close() error {
+	return newError(C.avio_close(ic.c))
+}
+
 func (ic *IOContext) Open(filename string, flags IOContextFlags) error {
 	cfi := C.CString(filename)
 	defer C.free(unsafe.Pointer(cfi))
 	return newError(C.avio_open(&ic.c, cfi, C.int(flags)))
 }
 
-func (ic *IOContext) Write(b []byte) {
+func (ic *IOContext) Write(b []byte) (int, error) {
 	if b == nil {
-		return
+		return 0, nil
 	}
 	C.avio_write(ic.c, (*C.uchar)(unsafe.Pointer(&b[0])), C.int(len(b)))
+	return len(b), nil
+}
+
+func (ic *IOContext) Read(b []byte) (n int, err error) {
+	n = int(C.avio_read(ic.c, (*C.uchar)(unsafe.Pointer(&b[0])), C.int(len(b))))
+	if n < 0 {
+		return 0, io.EOF
```

Comment:
I'd rather you return `0, newError(n)` instead

Reply:
i think `io.Reader` uses io.EOF so that is knows when to stop reading. Let me try it out though.

Reply:
Thing is, I want `astiav` to be as close as possible to C since it's merely an expression of the C code. Implementing `io.ReadWriteCloser` as well as the `io.Seeker` interfaces seems like a good idea but I'm wondering whether we should create a wrapper on top of `IOContext` instead 🤔 

Things that are bothering me:

- the `Close` method that would implement the `io.Closer` interface should use `Closep()` instead
- the `Write` method shouldn't return `(n int)` since C doesn't provide that information
- the `Read` method should return an `astiav` error

Here's what I'm proposing:

- update `IOContext` methods you've added so that they're closer to C (cf. my other comments)
- create an `IOContextWrapper` struct (let me know if you think of a better name) that would only implement the `io.ReadWriteCloser` as well as the `io.Seeker` interfaces (in those methods, you could return `io.EOF` for instance)
- add an `IOContext`.`Wrapper()` method that would return an `IOContextWrapper`

That people that want to stay close to C can use `IOContext`, and people that want to use GO interfaces can use `IOContextWrapper`

Reply:
Thanks for the proposal. I am currently looking at https://ffmpeg.org/doxygen/5.0/structAVIOContext.html

And it seems that there are some fields that I can use to extract information about how many bytes were written; testing in progress.

Concerning, the `io.Closer`, i actually just realized that we probably shouldn't implement it as `avio_closep(...)` (and `avio_close(...)`) is only needed when the resource is opened by`avio_open(...)`

I can manually check [eof_reached](https://ffmpeg.org/doxygen/5.0/structAVIOContext.html#aa9214f54c01f07076e0a0718b2e9c27b), and return `io.EOF` as the error.

What do you think? Do you still think a wrapper is better? I think the API will be confusing/weird with wrappers.

FYI: [Read needs to implement io.EOF](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/io/io.go;l=44)

Reply:
> I think the API will be confusing/weird with wrappers.

I agree with you.

I don't think implementing the `io.ReadWriteCloser` and `io.Seeker` is a good idea in this lib since its main mission is to be the closest possible to C. If somebody wants to implement this interface, he/she can do it in their project, it's pretty straight-forward.

Therefore here's what I'd rather have in this PR:

- add `func (ic *IOContext) Close() error`
- keep `Write()` as it is
- add `func (ic *IOContext) Read(b []byte) (n int, err error)`
- add `func (ic *IOContext) Seekable() bool`
- add `func (ic *IOContext) Seek(offset int64, whence int) (int64, error)`
- add `IOContext` getters if you need them to implement the `io` interfaces (such as `bytes_written` for instance)

and that's it. That should provide the proper method to anybody looking to implement `io.ReadWriteCloser` and `io.Seeker`.

Reply:
Makes sense will do.